### PR TITLE
Bugfix/typehints

### PR DESCRIPTION
**Pull request by dereuromark**
Follow up on https://github.com/cakephp/cakephp/pull/9426

IDE now reports less errors. Needs the previous PR to be merged first to show the actual commits on top only.

### CODE REVIEW

File: src/ORM/Query.php [hunk 1]
```diff
@@ -203,7 +202,7 @@ public function select($fields = [], $overwrite = false)
      *
      * This method returns the same query object for chaining.
      *
-     * @param \Cake\ORM\Table $table The table to pull types from
+     * @param \Cake\Datasource\RepositoryInterface $table The table to pull types from
      * @return $this
      */
     public function addDefaultTypes(Table $table)
```
Comment (dereuromark) [inline src/ORM/Query.php]: Is it reasonable to change this typehint to RepositoryInterface in 3.next, as well?

Comment (thinkingmedia) [inline src/ORM/Query.php]: But that interface doesn't have the methods that the function is using.

Comment (markstory) [inline src/ORM/Query.php]: Changing type hints doesn't allow for backwards compatibility as userland subclasses will trigger errors.

Comment (dereuromark) [inline src/ORM/Query.php]: @thinkingmedia Then we have a different problem here.
The typehint going in is of the interface type, so the IDE will complain that this is technically not quite correct, probably should be using more interface segregation in the future here and then we can always use interface typehints.

Anyway, I can then also revert the doc block change for this method and we keep the current warning.

**Comment by dereuromark** [comment]
Tests pass.

**Comment by lorenzo** [comment]
Thanks!

**Pull request merged**